src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -1969,6 +1970,30 @@ private void HandleVariableDeclaration(IVariableDeclarationOperation operation)
             {
                 ILocalSymbol localSymbol = declarator.Symbol;
 
+                // We skip constants in the control flow graph, as they're not actually involved in any control flow.
+                if (localSymbol.IsConst)
+                {
+                    continue;
+                }
+
+                // If the local is a static (possible in VB), then we create a semaphore for conditional execution of the initializer.
+                BasicBlock afterInitialization = null;
+                if (localSymbol.IsStatic && (declarator.Initializer != null || operation.Initializer != null))
+                {
+                    afterInitialization = new BasicBlock(BasicBlockKind.Block);
+                    LinkBlocks(CurrentBasicBlock, afterInitialization);
+
+                    ITypeSymbol booleanType = ((Operation)operation).SemanticModel.Compilation.GetSpecialType(SpecialType.System_Boolean);
+                    var initializationSemaphore = new StaticLocalInitializationSemaphoreOperation(localSymbol, localSymbol.DeclaringSyntaxReferences.Single().GetSyntax(), booleanType, constantValue: null, isImplicit: true);
+                    Operation.SetParentOperation(initializationSemaphore, null);
+
+                    var initializerBlock = new BasicBlock(BasicBlockKind.Block);
+                    LinkBlocks(CurrentBasicBlock, (initializationSemaphore, JumpIfTrue: true, RegularBranch(initializerBlock)));
+
+                    EnterRegion(new RegionBuilder(ControlFlowGraph.RegionKind.StaticLocalInitializer));
+                    AppendNewBlock(initializerBlock, linkToPrevious: false);
```

Comment:
>AppendNewBlock(initializerBlock, linkToPrevious: false); [](start = 20, length = 56)

Why ```linkToPrevious: false```? Also, I think the code would be simpler and more robust if we were not trying to allocate and add the first block in region explicitly, but instead let the machinery to do the allocation, addition and linking. In order to achieve that, we only link conditional branch above (this will also take care of double use of CurrentBasicBlock property), but we change its kind to JumpIfFalse and link it to afterInitialization. Then we simply enter the region and we are done, no need to allocate and add initializerBlock explicitly.

Reply:
We don't want to link to previous because there's no default branches to the initializer block. The only way to get to it is via the conditional set just above.

---
In reply to: [174864732](https://github.com/dotnet/roslyn/pull/25458#discussion_r174864732) [](ancestors = 174864732)

Reply:
Hmm. It struck me as more natural to have the conditional branch to be to the initializer, as I imagined the code would look more like this:
```
If (!initialized)
    ' Initialize
End If
```

I can give the inverted structure a try though.

---
In reply to: [174932845](https://github.com/dotnet/roslyn/pull/25458#discussion_r174932845) [](ancestors = 174932845,174864732)

Reply:
Changed to use this structure.

---
In reply to: [174933838](https://github.com/dotnet/roslyn/pull/25458#discussion_r174933838) [](ancestors = 174933838,174932845,174864732)